validator/client/src/main/java/tech/pegasys/artemis/validator/client/StableSubnetSubscriber.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.validator.client;
+
+import static java.lang.Integer.min;
+import static tech.pegasys.artemis.util.config.Constants.ATTESTATION_SUBNET_COUNT;
+import static tech.pegasys.artemis.util.config.Constants.EPOCHS_PER_RANDOM_SUBNET_SUBSCRIPTION;
+import static tech.pegasys.artemis.util.config.Constants.RANDOM_SUBNETS_PER_VALIDATOR;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import tech.pegasys.artemis.bls.BLSPublicKey;
+import tech.pegasys.artemis.validator.api.ValidatorApiChannel;
+
+public class StableSubnetSubscriber {
+
+  private final ValidatorApiChannel validatorApiChannel;
+  private final Map<BLSPublicKey, Validator> validators;
+  private final Set<Integer> availableSubnetIndices =
+      IntStream.range(0, ATTESTATION_SUBNET_COUNT).boxed().collect(Collectors.toSet());
+  private final Map<Integer, UnsignedLong> subnetIdToUnsubscriptionSlot = new HashMap<>();
```

Comment:
If you define a `SubnetSubscription` class, you can keep this as a `TreeSet` sorted by unsubscription slot.  Then we can find / replace expired subscriptions more efficiently.
```suggestion
  private final NavigableSet<SubnetSubscription> subnetSubscriptions = new TreeSet<>();
```

Reply:
Nice suggestion. Will do.

Reply:
Done.